cmd/metrics-v2.go

```diff
@@ -144,10 +144,75 @@ type Metric struct {
 	Histogram            map[string]uint64 `json:"Histogram"`
 }
 
+func (m *Metric) copyMetric() Metric {
+	metric := Metric{
+		Description:          m.Description,
+		Value:                m.Value,
+		HistogramBucketLabel: m.HistogramBucketLabel,
+		StaticLabels:         make(map[string]string),
+		VariableLabels:       make(map[string]string),
+		Histogram:            make(map[string]uint64),
+	}
+	for k, v := range m.StaticLabels {
+		metric.StaticLabels[k] = v
+	}
+	for k, v := range m.VariableLabels {
+		metric.VariableLabels[k] = v
+	}
+	for k, v := range m.Histogram {
+		metric.Histogram[k] = v
+	}
+	return metric
+}
+
+// metricsGroupCacheEntry captures the cached metrics for a metrics group.
+type metricsGroupCacheEntry struct {
+	lastUpdate time.Time
+	cache      []Metric
+}
+
 // MetricsGroup are a group of metrics that are initialized together.
 type MetricsGroup struct {
-	Metrics    []Metric
-	initialize func(ctx context.Context, m *MetricsGroup)
+	id            string
+	Metrics       []Metric
+	cacheInterval time.Duration
+	cachedRead    func(ctx context.Context, m *MetricsGroup)
+	read          func(ctx context.Context, m *MetricsGroup)
+}
+
+var metricsGroupCache = make(map[string]metricsGroupCacheEntry)
+var cacheLock sync.Mutex
+
+func cachedRead(ctx context.Context, m *MetricsGroup) {
+	cacheLock.Lock()
+	defer cacheLock.Unlock()
+	readMetrics := func() {
+		m.read(ctx, m)
+		c := make([]Metric, 0)
+		for i := range m.Metrics {
+			c = append(c, m.Metrics[i].copyMetric())
+		}
+		metricsGroupCache[m.id] = metricsGroupCacheEntry{
+			lastUpdate: time.Now(),
+			cache:      c,
+		}
+	}
+	c, ok := metricsGroupCache[m.id]
+	if !ok || m.cacheInterval == -1 {
+		readMetrics()
+		return
+	}
+	interval := m.cacheInterval
+	if interval == 0 {
+		interval = time.Second * 30
+	}
+	if len(c.cache) > 0 && time.Since(c.lastUpdate) < interval {
+		for i := range c.cache {
+			m.Metrics = append(m.Metrics, c.cache[i].copyMetric())
+		}
+		return
+	}
+	readMetrics()
```

Comment:
This type of code is not necessary you can re-implement this using timedValue data structure that can cache just the way its written here and its already widely used @kerneltime

Reply:
Updated.